### PR TITLE
Avoid overflow in ThreadSet::any() and nits

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

I missed to spot one edge-case overflow bug at https://github.com/solana-labs/solana/pull/30422 while reviewing.

#### Summary of Changes

Fix it along with really tiny nits.

Fixes #

### CODE REVIEW

File: core/src/banking_stage/thread_aware_account_locks.rs [hunk 1]
```diff
@@ -310,12 +308,16 @@ impl Debug for ThreadSet {
...
-        Self(0)
+        Self(0b0)
     }
 
     #[inline(always)]
     pub(crate) const fn any(num_threads: usize) -> Self {
-        Self(Self::as_flag(num_threads) - 1)
+        if num_threads == MAX_THREADS {
+            Self(u64::MAX)
+        } else {
```
Comment (ryoqun) [inline core/src/banking_stage/thread_aware_account_locks.rs:318]: obviously, playing dumb regarding `num_threads > MAX_THREADS` as was before.. xD

File: core/src/banking_stage/thread_aware_account_locks.rs [hunk 2]
```diff
@@ -310,12 +308,16 @@ impl Debug for ThreadSet {
...
     #[inline(always)]
     pub(crate) const fn none() -> Self {
-        Self(0)
+        Self(0b0)
     }
 
     #[inline(always)]
     pub(crate) const fn any(num_threads: usize) -> Self {
-        Self(Self::as_flag(num_threads) - 1)
+        if num_threads == MAX_THREADS {
```
Comment (ryoqun) [inline core/src/banking_stage/thread_aware_account_locks.rs:316]: hate to introduce branches in this `#[inline(always)]`-ed code... but this is a compromised code result, struggling for cleaner code.

**Review by apfitzge — APPROVED** [review]
lgtm - nice catches!

File: core/src/banking_stage/thread_aware_account_locks.rs [hunk 3]
```diff
@@ -310,12 +308,16 @@ impl Debug for ThreadSet {
 impl ThreadSet {
     #[inline(always)]
     pub(crate) const fn none() -> Self {
-        Self(0)
+        Self(0b0)
     }
 
     #[inline(always)]
     pub(crate) const fn any(num_threads: usize) -> Self {
-        Self(Self::as_flag(num_threads) - 1)
```
Comment (ryoqun) [inline core/src/banking_stage/thread_aware_account_locks.rs:318]: under debug build, this causes following overflow panic:

```
---- banking_stage::thread_aware_account_locks::tests::test_thread_set_any_max stdout ----
thread 'banking_stage::thread_aware_account_locks::tests::test_thread_set_any_max' panicked at 'attempt to shift left with overflow', core/src/banking_stage/thread_aware_account_locks.rs:361:9
stack backtrace:
   0:     0x564463f5f4da - std::backtrace_rs::backtrace::libunwind::trace::h6ee169f87c9eb811
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x564463f5f4da - std::backtrace_rs::backtrace::trace_unsynchronized::hae1aa8caf0f73946
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x564463f5f4da - std::sys_common::backtrace::_print_fmt::hd0ae816e159c88c0
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x564463f5f4da - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h73a43b159f05d7bf
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x564463f8acce - core::fmt::write::h098923995eefaba9
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/core/src/fmt/mod.rs:1213:17
   5:     0x564463f56ec5 - std::io::Write::write_fmt::h38053d35a74cabc0
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/io/mod.rs:1682:15
   6:     0x564463f5f2a5 - std::sys_common::backtrace::_print::h1dd4ca615e620472
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x564463f5f2a5 - std::sys_common::backtrace::print::hdaad52d464d17c88
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x564463f6109f - std::panicking::default_hook::{{closure}}::h1c9cf4fd4a868b33
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/panicking.rs:267:22
   9:     0x564463f60d5f - std::panicking::default_hook::h3725ab7a4be0fc2f
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/std/src/panicking.rs:283:9
  10:     0x5644609935a7 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h94186f52c5d088a0
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/alloc/src/boxed.rs:2002:9
  11:     0x5644609935a7 - test::test_main::{{closure}}::h485d089d9dba2c03
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/test/src/lib.rs:135:21
  12:     0x564463f6182a - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h3f090c481fefe13f
                               at /rustc/52372f9c71d8ade4cb815524f179119656f0aa2e/library/alloc/src/boxed.rs:2002:9
  13:     0x564463f6182a - std::panicking::rust_panic_with_hook::hb7593c6583c1a088
```

on the other hand, it should return correct result under release build, where overflow checks aren't done. out of pure luck, that leads to correct behavior in our case.

File: core/src/banking_stage/thread_aware_account_locks.rs [hunk 4]
```diff
@@ -310,12 +308,16 @@ impl Debug for ThreadSet {
...
+        if num_threads == MAX_THREADS {
+            Self(u64::MAX)
```
Comment (ryoqun) [inline core/src/banking_stage/thread_aware_account_locks.rs:317]: btw, i also considered `Self(!0b0)` for type agnotic code, but chose the const albeit, while being a bit off-meaning here... lol

**Pull request merged**